File: api/src/main/java/org/apache/gravitino/semantic/Relationship.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.gravitino.semantic;

import com.google.common.base.Preconditions;
import java.util.Arrays;
import java.util.Objects;
import javax.annotation.Nullable;
import org.apache.gravitino.annotation.Evolving;

/**
 * An immutable relationship between two datasets in the same semantic model. The endpoint column
 * arrays describe the corresponding join columns on the source and target datasets.
 */
@Evolving
public final class Relationship {

  private final String name;
  private final String from;
  private final String to;
  private final String[] fromColumns;
  private final String[] toColumns;

  @Nullable private final AIContext aiContext;
  @Nullable private final CustomExtension[] customExtensions;

  private Relationship(Builder builder) {
    this.name = builder.name;
    this.from = builder.from;
    this.to = builder.to;
    this.fromColumns = Arrays.copyOf(builder.fromColumns, builder.fromColumns.length);
    this.toColumns = Arrays.copyOf(builder.toColumns, builder.toColumns.length);
    this.aiContext = builder.aiContext;
    this.customExtensions =
        builder.customExtensions == null
            ? null
            : Arrays.copyOf(builder.customExtensions, builder.customExtensions.length);
  }

  /**
   * Creates a builder for an immutable {@link Relationship}.
   *
   * @return A new builder.
   */
  public static Builder builder() {
    return new Builder();
  }

  /**
   * Returns the relationship name.
   *
   * @return The relationship name.
   */
  public String name() {
    return name;
  }

  /**
   * Returns the name of the source dataset.
   *
   * @return The source dataset name.
   */
  public String from() {
    return from;
  }

  /**
   * Returns the name of the target dataset.
   *
   * @return The target dataset name.
   */
  public String to() {
    return to;
  }

  /**
   * Returns a copy of the source dataset columns.
   *
   * @return The source columns.
   */
  public String[] fromColumns() {
    return Arrays.copyOf(fromColumns, fromColumns.length);
  }

  /**
   * Returns a copy of the target dataset columns.
   *
   * @return The target columns.
   */
  public String[] toColumns() {
    return Arrays.copyOf(toColumns, toColumns.length);
  }

  /**
   * Returns the AI context associated with the relationship.
   *
   * @return The AI context, or {@code null} if it is not set.
   */
  @Nullable
  public AIContext aiContext() {
    return aiContext;
  }

  /**
   * Returns a copy of the custom extensions associated with the relationship.
   *
   * @return The custom extensions, or {@code null} if they are not set.
   */
  @Nullable
  public CustomExtension[] customExtensions() {
    return customExtensions == null
        ? null
        : Arrays.copyOf(customExtensions, customExtensions.length);
  }

  /**
   * Compares this relationship with another object for value equality.
   *
   * @param other The object to compare with.
   * @return {@code true} if the objects are equal, otherwise {@code false}.
   */
  @Override
  public boolean equals(@Nullable Object other) {
    if (this == other) {
      return true;
    }
    if (!(other instanceof Relationship)) {
      return false;
    }
    Relationship that = (Relationship) other;
    return name.equals(that.name)
        && from.equals(that.from)
        && to.equals(that.to)
        && Arrays.equals(fromColumns, that.fromColumns)
        && Arrays.equals(toColumns, that.toColumns)
        && Objects.equals(aiContext, that.aiContext)
        && Arrays.equals(customExtensions, that.customExtensions);
  }

  /**
   * Returns the value-based hash code for this relationship.
   *
   * @return The hash code.
   */
  @Override
  public int hashCode() {
    return Objects.hash(
        name,
        from,
        to,
        Arrays.hashCode(fromColumns),
        Arrays.hashCode(toColumns),
        aiContext,
        Arrays.hashCode(customExtensions));
  }

  /**
   * Returns a string representation of this relationship.
   *
   * @return The string representation.
   */
  @Override
  public String toString() {
    return "Relationship{"
        + "name='"
        + name
        + '\''
        + ", from='"
        + from
        + '\''
        + ", to='"
        + to
        + '\''
        + ", fromColumns="
        + Arrays.toString(fromColumns)
        + ", toColumns="
        + Arrays.toString(toColumns)
        + ", aiContext="
        + aiContext
        + ", customExtensions="
        + Arrays.toString(customExtensions)
        + '}';
  }

  /** A builder for immutable {@link Relationship} values. */
  public static final class Builder {

    private String name;
    private String from;
    private String to;
    private String[] fromColumns;
    private String[] toColumns;

    @Nullable private AIContext aiContext;
    @Nullable private CustomExtension[] customExtensions;

    private Builder() {}

    /**
     * Sets the relationship name.
     *
     * @param name The non-empty relationship name.
     * @return This builder.
     */
    public Builder withName(String name) {
      this.name = name;
      return this;
    }

    /**
     * Sets the source dataset name.
     *
     * @param from The non-empty source dataset name.
     * @return This builder.
     */
    public Builder withFrom(String from) {
      this.from = from;
      return this;
    }

    /**
     * Sets the target dataset name.
     *
     * @param to The non-empty target dataset name.
     * @return This builder.
     */
    public Builder withTo(String to) {
      this.to = to;
      return this;
    }

    /**
     * Sets the source dataset columns.
     *
     * @param fromColumns The non-empty source column array.
     * @return This builder.
     */
    public Builder withFromColumns(String[] fromColumns) {
      this.fromColumns = fromColumns;
      return this;
    }

    /**
     * Sets the target dataset columns.
     *
     * @param toColumns The non-empty target column array.
     * @return This builder.
     */
    public Builder withToColumns(String[] toColumns) {
      this.toColumns = toColumns;
      return this;
    }

    /**
     * Sets the optional AI context.
     *
     * @param aiContext The AI context, or {@code null} to leave it unset.
     * @return This builder.
     */
    public Builder withAIContext(@Nullable AIContext aiContext) {
      this.aiContext = aiContext;
      return this;
    }

    /**
     * Sets the optional custom extensions.
     *
     * @param customExtensions The custom extensions, or {@code null} to leave them unset.
     * @return This builder.
     */
    public Builder withCustomExtensions(@Nullable CustomExtension[] customExtensions) {
      this.customExtensions = customExtensions;
      return this;
    }

    /**
     * Builds an immutable {@link Relationship}.
     *
     * @return The new relationship.
     * @throws IllegalArgumentException If a required name is null or empty, a required column array
     *     is null or empty, a column name or custom extension is null, or the endpoint column
     *     arrays have different lengths.
     */
    public Relationship build() {
      Preconditions.checkArgument(
          name != null && !name.isEmpty(), "name must not be null or empty");
      Preconditions.checkArgument(
          from != null && !from.isEmpty(), "from must not be null or empty");
      Preconditions.checkArgument(to != null && !to.isEmpty(), "to must not be null or empty");
      Preconditions.checkArgument(
          fromColumns != null && fromColumns.length > 0, "fromColumns must not be null or empty");
      Preconditions.checkArgument(
          toColumns != null && toColumns.length > 0, "toColumns must not be null or empty");
      validateColumnNames("fromColumns", fromColumns);
      validateColumnNames("toColumns", toColumns);
      Preconditions.checkArgument(
          fromColumns.length == toColumns.length,
          "fromColumns and toColumns must have the same length");
      Preconditions.checkArgument(
          customExtensions == null || Arrays.stream(customExtensions).allMatch(Objects::nonNull),
          "customExtensions must not contain null");
      return new Relationship(this);
    }
  }

  private static void validateColumnNames(String name, String[] columns) {
    for (int index = 0; index < columns.length; index++) {
      Preconditions.checkArgument(
          columns[index] != null && !columns[index].isEmpty(),
          "%s[%s] must not be null or empty",
          name,
          index);
    }
  }
}

File: clients/client-python/gravitino/api/semantic/relationship.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from typing import Optional

from gravitino.api.semantic.ai_context import AIContext
from gravitino.api.semantic.custom_extension import CustomExtension
from gravitino.api.semantic.semantic_utils import (
    check_no_none_elements,
    check_non_empty_string_elements,
)
from gravitino.utils.precondition import Precondition


class Relationship:  # pylint: disable=too-many-instance-attributes
    """A join between two datasets in the same Semantic Model.

    Relationship names are unique within a Semantic Model. Both endpoints name a
    dataset in the same Semantic Model, and the joined column lists are non-empty
    and of equal length.

    ``from_dataset`` and ``to_dataset`` map to the Ossie ``from`` and ``to``
    fields, which cannot be used as Python identifiers.
    """

    def __init__(
        self,
        name: str,
        from_dataset: str,
        to_dataset: str,
        from_columns: list[str],
        to_columns: list[str],
        ai_context: Optional[AIContext] = None,
        custom_extensions: Optional[list[CustomExtension]] = None,
    ):
        Precondition.check_argument(
            name is not None and name != "", "name must not be null or empty"
        )
        Precondition.check_argument(
            from_dataset is not None and from_dataset != "",
            "from must not be null or empty",
        )
        Precondition.check_argument(
            to_dataset is not None and to_dataset != "", "to must not be null or empty"
        )
        Precondition.check_argument(
            from_columns is not None and len(from_columns) > 0,
            "fromColumns must not be null or empty",
        )
        Precondition.check_argument(
            to_columns is not None and len(to_columns) > 0,
            "toColumns must not be null or empty",
        )
        check_non_empty_string_elements("fromColumns", from_columns)
        check_non_empty_string_elements("toColumns", to_columns)
        Precondition.check_argument(
            len(from_columns) == len(to_columns),
            "fromColumns and toColumns must have the same length",
        )
        check_no_none_elements("customExtensions", custom_extensions)

        self._name = name
        self._from_dataset = from_dataset
        self._to_dataset = to_dataset
        self._from_columns = list(from_columns)
        self._to_columns = list(to_columns)
        self._ai_context = ai_context
        self._custom_extensions = (
            None if custom_extensions is None else list(custom_extensions)
        )

    def name(self) -> str:
        """Returns the relationship name."""
        return self._name

    def from_dataset(self) -> str:
        """Returns the name of the dataset the relationship joins from."""
        return self._from_dataset

    def to_dataset(self) -> str:
        """Returns the name of the dataset the relationship joins to."""
        return self._to_dataset

    def from_columns(self) -> list[str]:
        """Returns the joined columns exposed by the source dataset."""
        return list(self._from_columns)

    def to_columns(self) -> list[str]:
        """Returns the joined columns exposed by the target dataset."""
        return list(self._to_columns)

    def ai_context(self) -> Optional[AIContext]:
        """Returns the AI context, or `None` if it is not set."""
        return self._ai_context

    def custom_extensions(self) -> Optional[list[CustomExtension]]:
        """Returns the custom extensions, or `None` if they are not set."""
        return (
            None if self._custom_extensions is None else list(self._custom_extensions)
        )

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Relationship):
            return False
        return (
            self._name == other.name()
            and self._from_dataset == other.from_dataset()
            and self._to_dataset == other.to_dataset()
            and self._from_columns == other.from_columns()
            and self._to_columns == other.to_columns()
            and self._ai_context == other.ai_context()
            and self._custom_extensions == other.custom_extensions()
        )

    def __hash__(self) -> int:
        return hash(
            (
                self._name,
                self._from_dataset,
                self._to_dataset,
                tuple(self._from_columns),
                tuple(self._to_columns),
                self._ai_context,
                (
                    None
                    if self._custom_extensions is None
                    else tuple(self._custom_extensions)
                ),
            )
        )

    def __repr__(self) -> str:
        return (
            f"Relationship(name={self._name!r}, from={self._from_dataset!r}, "
            f"to={self._to_dataset!r}, fromColumns={self._from_columns!r}, "
            f"toColumns={self._to_columns!r}, aiContext={self._ai_context!r}, "
            f"customExtensions={self._custom_extensions!r})"
        )
